signac/contrib/job.py

```diff
@@ -88,29 +229,27 @@ class Job:
     def __init__(self, project, statepoint=None, _id=None):
         self._project = project
 
+        # Prepare wd in advance so that the attribute exists in checks below.
+        self._wd = None
+
         if statepoint is None and _id is None:
             raise ValueError("Either statepoint or _id must be provided.")
         elif statepoint is not None:
-            # A state point was provided.
-            self._statepoint = SyncedAttrDict(statepoint, parent=_sp_save_hook(self))
-            # If the id is provided, assume the job is already registered in
-            # the project cache and that the id is valid for the state point.
-            if _id is None:
-                # Validate the state point and recursively convert to supported types.
-                statepoint = self.statepoint()
-                # Compute the id from the state point if not provided.
-                self._id = calc_id(statepoint)
-                # Update the project's state point cache immediately if opened by state point
-                self._project._register(self.id, statepoint)
-            else:
-                self._id = _id
+            self._statepoint_requires_init = False
+            self._id = calc_id(statepoint) if _id is None else _id
+            self._statepoint = _StatepointDict(
+                jobs=[self], filename=self._statepoint_filename, data=statepoint
```

Comment:
How does shallow-copying a Job trigger an append to the _jobs property of the _StatepointDict? (I started discussing with @vyasr on Slack but we're deferring this conversation for now until one of us can investigate.)

Reply:
After a copy is made `__setstate__` will get called, which is where the list is appended to.